### PR TITLE
Adding Run Information row in the Result Email alert

### PR DESCRIPTION
**Pull request by pdhiran**
<s>Suspecting that this commit is causing the tests to send Not Run status report. Will debug and fix the issue</s>.
Issue with the Openstack connection, due to which tests were failing during node creation. Added a Run Information column information row in the email, which provides information about failure during node creation if any, otherwise gives the run status.

Signed-off-by: Pawan Dhiran <pdhiran@redhat.com>

# Description

Please include Automation development guidelines. Source of Test case - New Feature/Regression Test/Close loop of customer BZs

# Checklist:

- [ ] Create a test case in Polarion reviewed and approved.
- [ ] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [ ] Review the automation design
- [ ] Implement the test script and perform test runs
- [ ] Submit PR for code review and approve
- [ ] Update Polarin Test with Automation script details and update automation fields
- [ ] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test

### CODE REVIEW

**Comment by sunilkumarn417** [comment]
@pdhiran can you add some more information on issue.

**Comment by pdhiran** [comment]
For the tests triggered via Jenkins, in the result email, the log link at magna gave 404 not found error. Looks like issue with mount on the jenkins slave. I thought my PR broke the logging. Also there was a connection outage with openstack. 
``` 2021-01-15 16:06:10,405 - ceph.ceph - WARNING - Connection outage:
('Bad authentication type', ['publickey', 'gssapi-keyex', 'gssapi-with-mic']) (allowed_types=['publickey', 'gssapi-keyex', 'gssapi-with-mic'])
2021-01-15 16:06:20,408 - paramiko.transport - ERROR - Exception: Error reading SSH protocol banner
2021-01-15 16:06:20,408 - paramiko.transport - ERROR - Traceback (most recent call last):
2021-01-15 16:06:20,409 - paramiko.transport - ERROR -   File "/home/pdhiran/new/cephci/lib64/python3.6/site-packages/paramiko/transport.py", line 2138, in _check_banner
2021-01-15 16:06:20,409 - paramiko.transport - ERROR -     buf = self.packetizer.readline(timeout)
2021-01-15 16:06:20,409 - paramiko.transport - ERROR -   File "/home/pdhiran/new/cephci/lib64/python3.6/site-packages/paramiko/packet.py", line 367, in readline
2021-01-15 16:06:20,409 - paramiko.transport - ERROR -     buf += self._read_timeout(timeout)
2021-01-15 16:06:20,409 - paramiko.transport - ERROR -   File "/home/pdhiran/new/cephci/lib64/python3.6/site-packages/paramiko/packet.py", line 563, in _read_timeout
2021-01-15 16:06:20,409 - paramiko.transport - ERROR -     raise EOFError()
2021-01-15 16:06:20,409 - paramiko.transport - ERROR - EOFError
2021-01-15 16:06:20,409 - paramiko.transport - ERROR -
2021-01-15 16:06:20,409 - paramiko.transport - ERROR - During handling of the above exception, another exception occurred:
2021-01-15 16:06:20,409 - paramiko.transport - ERROR -
2021-01-15 16:06:20,409 - paramiko.transport - ERROR - Traceback (most recent call last):
2021-01-15 16:06:20,409 - paramiko.transport - ERROR -   File "/home/pdhiran/new/cephci/lib64/python3.6/site-packages/paramiko/transport.py", line 1966, in run
2021-01-15 16:06:20,409 - paramiko.transport - ERROR -     self._check_banner()
2021-01-15 16:06:20,409 - paramiko.transport - ERROR -   File "/home/pdhiran/new/cephci/lib64/python3.6/site-packages/paramiko/transport.py", line 2143, in _check_banner
2021-01-15 16:06:20,409 - paramiko.transport - ERROR -     "Error reading SSH protocol banner" + str(e)
2021-01-15 16:06:20,409 - paramiko.transport - ERROR - paramiko.ssh_exception.SSHException: Error reading SSH protocol banner
2021-01-15 16:06:20,410 - paramiko.transport - ERROR -
2021-01-15 16:06:20,410 - ceph.ceph - WARNING - Connection outage:
Error reading SSH protocol banner
2021-01-15 16:06:30,423 - paramiko.transport - INFO - Connected (version 2.0, client OpenSSH_7.4)
2021-01-15 16:06:30,477 - ceph.ceph - WARNING - Connection outage:
('Bad authentication type', ['publickey', 'gssapi-keyex', 'gssapi-with-mic']) (allowed_types=['publickey', 'gssapi-keyex', 'gssapi-with-mic'])
2021-01-15 16:06:40,490 - paramiko.transport - INFO - Connected (version 2.0, client OpenSSH_7.4)
2021-01-15 16:06:40,542 - ceph.ceph - WARNING - Connection outage:
('Bad authentication type', ['publickey', 'gssapi-keyex', 'gssapi-with-mic']) (allowed_types=['publickey', 'gssapi-keyex', 'gssapi-with-mic'])
2021-01-15 16:06:50,556 - paramiko.transport - INFO - Connected (version 2.0, client OpenSSH_7.4)
2021-01-15 16:06:50,607 - ceph.ceph - WARNING - Connection outage:
('Bad authentication type', ['publickey', 'gssapi-keyex', 'gssapi-with-mic']) (allowed_types=['publickey', 'gssapi-keyex', 'gssapi-with-mic'])
2021-01-15 16:06:50,608 - __main__ - ERROR - ('Bad authentication type', ['publickey', 'gssapi-keyex', 'gssapi-with-mic']) (allowed_types=['publickey', 'gssapi-keyex', 'gssapi-with-mic'])
2021-01-15 16:06:50,653 - utility.utils - INFO - Results have been emailed to ['pdhiran@redhat.com']
Traceback (most recent call last):
  File "run.py", line 660, in <module>
    log.info("final rc of test run %d" % rc)
TypeError: %d format: a number is required, not NoneType ```

**Comment by sunilkumarn417** [comment]
I think returning `rc` value should fix the issue, since in main its expecting the value. However broad exception handling should handle all the errors at `create_nodes`.

```
    if reuse is None:
        try:
            ceph_cluster_dict, clients = create_nodes(conf, inventory, osp_cred, run_id, service, instances_name)
        except Exception as err:
            log.error(err)
            tests = suite.get('tests')
            res = []
            for test in tests:
                test = test.get('test')
                tmp = fetch_test_details(test)
                res.append(tmp)
            total_time = "0s"
            send_to_cephci = post_results or post_to_report_portal
            email_results(res, run_id, trigger_user, run_dir, total_time, send_to_cephci)
            return 1

.....
....
if __name__ == '__main__':
    args = docopt(doc)
    rc = run(args)
    log.info("final rc of test run %d" % rc)
    sys.exit(rc)
```

**Comment by pdhiran** [comment]
Test Run: http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-1610800769345/

File: run.py [hunk 1]
```diff
@@ -452,8 +452,9 @@ def fetch_test_details(var) -> dict:
                 res.append(tmp)
             total_time = "0s"
             send_to_cephci = post_results or post_to_report_portal
-            email_results(res, run_id, trigger_user, run_dir, total_time, send_to_cephci)
-            return
+            info = f"Exception hit during node creation.Traceback: {traceback.format_exc(limit=2)}"
```
Comment (sunilkumarn417) [inline run.py:455]: Can we keep this `environment setup` status attribute in html page and need `success` or `OKAY` for successful environment setup.

File: run.py [hunk 2]
```diff
@@ -452,8 +452,9 @@ def fetch_test_details(var) -> dict:
...
+            email_results(res, run_id, trigger_user, run_dir, total_time, info, send_to_cephci)
```
Comment (sunilkumarn417) [inline run.py:456]: would suggest to apply *args, **kwargs if it is helpful.

File: run.py (hunk 1 above)
Comment (sunilkumarn417) [reply run.py:455]: if it is node creation failure, it would be easy to address as environment failure. I think it may be not that easy with trace-back failure to say whether its environment or test failure straightaway.

File: run.py [hunk 3]
```diff
@@ -483,6 +484,7 @@ def fetch_test_details(var) -> dict:
     tests = suite.get('tests')
     tcs = []
     jenkins_rc = 0
+    fail_count = 0
```
Comment (sunilkumarn417) [inline run.py:487]: Hope this may not be needed, since we already receiving `PASSED` or `FAILED` in the email header. 
But I think we may need another one `SETUP_ERROR` which indicates environment issue. what do you think?

File: run.py [hunk 4]
```diff
@@ -599,6 +601,7 @@ def fetch_test_details(var) -> dict:
                 break
         elapsed = (datetime.datetime.now() - start)
         tc['duration'] = elapsed
+        test_info = "All tests in the suite have passed. Run Successful!"
```
Comment (sunilkumarn417) [inline run.py:604]: addressed in the previous comment.

File: utility/utils.py [hunk 5]
```diff
@@ -533,7 +533,7 @@ def custom_ceph_config(suite_config, custom_config, custom_config_file):
     return full_custom_config
 
 
-def email_results(results_list, run_id, trigger_user, run_dir, suite_run_time, send_to_cephci=False):
+def email_results(results_list, run_id, trigger_user, run_dir, suite_run_time, info, send_to_cephci=False):
```
Comment (sunilkumarn417) [inline utility/utils.py:536]: *args, **kwargs, addressed same in the previous comment in calling module.

**Review by sunilkumarn417 — CHANGES_REQUESTED** [review]

File: run.py (hunk 1 above)
Comment (rakeshgm) [reply run.py:455]: yeah, we can't have traceback information in the emails. just add the status. use proper flag.

File: run.py (hunk 3 above)
Comment (rakeshgm) [reply run.py:487]: so you mean `SETUP_ERROR` in the mail header. that's sound right. but it would also be nice have the failed or passed count in the summary.

File: run.py [hunk 6]
```diff
@@ -643,7 +648,7 @@ def fetch_test_details(var) -> dict:
     total_run_time = (datetime.datetime.now() - run_start_time)
     total_time = """'''Start Time:''' {a} '''End Time:''' {b} '''Total duration:'''{c} \
```
Comment (rakeshgm) [inline run.py:649]: while this is not related to this PR. correcting this would be better for readability. spit into different rows and add sub rows or columns in this one itself.

File: run.py (hunk 3 above)
Comment (sunilkumarn417) [reply run.py:487]: Yes, or like `ABORTED` or `SETUP_FAILURE`.

File: run.py (hunk 6 above)
Comment (rakeshgm) [reply run.py:649]: but you can choose to dismiss this if you want to take this in a separate PR

File: run.py (hunk 2 above)
Comment (psathyan) [reply run.py:456]: It is a good direction. IMO, instead of keeping it generic, we could use `test_results` or any other variable name of type `dict`. This would allow us to add/modify without changing the method signature.

File: run.py [hunk 7]
```diff
@@ -452,8 +452,9 @@ def fetch_test_details(var) -> dict:
                 res.append(tmp)
```
Comment (psathyan) [inline run.py:452]: We are discovering the tests again further down. It would be nice if we do it once... I like the idea of discovering the tests before the execution `setup | run | teardown `

File: utility/utils.py (hunk 5 above)
Comment (psathyan) [reply utility/utils.py:536]: IMO, the argument could be a dictionary having the required and/or optional keys. It would address Sunil's concern of having it flexible. This would require proper docstring though...

File: utility/utils.py [hunk 8]
```diff
@@ -580,7 +581,7 @@ def email_results(results_list, run_id, trigger_user, run_dir, suite_run_time, s
         msg['From'] = sender
         msg['To'] = ", ".join(recipients)
         html = create_html_file(run_name=run_name, results_list=results_list, run_time=suite_run_time,
-                                user=trigger_user, run_dir=run_dir)
+                                user=trigger_user, run_dir=run_dir, info=info)
```
Comment (psathyan) [inline utility/utils.py:584]: As the test result data is being used in multiple places, it might require a separate PR align the movement.

File: run.py (hunk 1 above)
Comment (pdhiran) [reply run.py]: Made the suggested changes.

File: run.py (hunk 2 above)
Comment (pdhiran) [reply run.py]: Added a dictionary to pass various variables to the email alert method.

File: run.py (hunk 3 above)
Comment (pdhiran) [reply run.py]: Added "SETUP_FAILURE" in the email header if there is a error in environment setup.

File: run.py (hunk 6 above)
Comment (pdhiran) [reply run.py]: Added a separate section with the timing details

File: run.py (hunk 7 above)
Comment (pdhiran) [reply run.py:452]: Will handle this in another PR.

File: utility/utils.py [hunk 9]
```diff
@@ -533,20 +534,23 @@ def custom_ceph_config(suite_config, custom_config, custom_config_file):
     return full_custom_config
 
 
-def email_results(results_list, run_id, trigger_user, run_dir, suite_run_time, send_to_cephci=False):
+def email_results(email_result):
```
Comment (psathyan) [inline utility/utils.py:537]: IMO, `email_result` should be `test_result` or `test_data` as it contains information about the tests/suites...

File: utility/utils.py [hunk 10]
```diff
@@ -557,6 +561,16 @@ def email_results(results_list, run_id, trigger_user, run_dir, suite_run_time, s
...
+    send_to_cephci = email_result.get("send_to_cephci", False)
+    try:
+        run_id = email_result["run_id"]
+        results_list = email_result["result"]
+        trigger_user = email_result["trigger_user"]
+        run_dir = email_result["run_directory"]
+        suite_run_time = email_result['total_time']
+        info = email_result['info']
+    except KeyError as kerr:
+        log.error(f"Key not found : {kerr}")
```
Comment (psathyan) [inline utility/utils.py:573]: Is it `fatal`? If yes, please exit and don't proceed further due to lack of information. Alternatively, we can chose to use `.get` if it is acceptable not to have those values.

File: utility/utils.py [hunk 11]
```diff
@@ -602,33 +615,37 @@ def email_results(results_list, run_id, trigger_user, run_dir, suite_run_time, s
...
         user (str): user of the node where the run is triggered from
         run_dir (str): log directory path
-        run_time (str): suite total duration info
+        run_time (dict): suite total duration info
+        info (dict): General information about the test run
     Returns: HTML file
 
     """
     log_link = "http://magna002.ceph.redhat.com/cephci-jenkins/{run}/".format(run=run_name)
+    info["link"] = f"{log_link}/startup.log"
```
Comment (psathyan) [inline utility/utils.py:632]: Let's keep the hyperlinks `relative`. It facilities... moving to different locations in case of local runs.

File: templates/result-email-template.html [hunk 12]
```diff
@@ -72,13 +98,23 @@ <h2>General Information</h2>
...
+            <td>{{ suite_run_time.start }}</td>
         </tr>
+        <tr>
+            <th style="color: blue">End Time</th>
+            <td>{{ suite_run_time.end }}</td>
         </tr>
         <tr>
-            <th style="color: blue">Invoked by User</th>
-            <td>{{ trigger_user }}</td>
+            <th style="color: blue">Total Run Time</th>
```
Comment (psathyan) [inline templates/result-email-template.html:116]: Let's keep it as `Duration` | `Execution Duration`

File: utility/utils.py [hunk 13]
```diff
@@ -602,33 +615,37 @@ def email_results(results_list, run_id, trigger_user, run_dir, suite_run_time, s
...
     Creates the HTML file from the template to be sent via mail
     Args:
         results_list (list): test case results info
         run_name (str): name of the test run
         user (str): user of the node where the run is triggered from
         run_dir (str): log directory path
-        run_time (str): suite total duration info
+        run_time (dict): suite total duration info
+        info (dict): General information about the test run
     Returns: HTML file
```
Comment (psathyan) [inline utility/utils.py:629]: Kindly remove this blank line

File: templates/result-email-template.html (hunk 12 above)
Comment (pdhiran) [reply templates/result-email-template.html]: Changed to Execution Duration

File: utility/utils.py (hunk 10 above)
Comment (pdhiran) [reply utility/utils.py:567]: added a exit call with 1

**Comment by pdhiran** [comment]
Not Run Status : http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-1611130270791/
Suite Pass Status : http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-1611126587125/
Suite Failure status : http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-1611121046787/

**Review by psathyan — COMMENTED** [review]
Would it be possible to move the `Environment Setup` details above the `Test Results`? At present, it is the first table. The title could be `Suite Setup`... IMO, the order could be
- General Information
- Suite Setup
- Test Results

Change `General Information` to `Summary`. Here, please merge the `suite duration`. Alternatively, `| Execution Summary | Start time : {time} | End time: {time} | Duration: {time} |`

The duration value looks in the report of `http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-1611121046787/result.html`. I hope that is the final rendering.

File: setup.py [hunk 14]
```diff
@@ -21,8 +21,9 @@
         'requests==2.21.0',
         'paramiko==2.4.2',
         'pyyaml>=4.2b1',
-        'jinja2==2.10.1',
-        'junitparser==1.4.0'
+        'jinja2',
```
Comment (rakeshgm) [inline setup.py:24]: why did you remove the version number here ?

**Comment by psathyan** [comment]
Nice work... the duration looks good at http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-1611121046787/result.html

File: templates/result-email-template.html [hunk 15]
```diff
@@ -72,13 +98,23 @@ <h2>General Information</h2>
...
         </tr>
+        <tr>
+            <th style="color: blue">End Time</th>
+            <td>{{ suite_run_time.end }}</td>
         </tr>
         <tr>
-            <th style="color: blue">Invoked by User</th>
-            <td>{{ trigger_user }}</td>
+            <th style="color: blue">Execution Duration</th>
+            <td>{{ suite_run_time.total }}</td>
```
Comment (rakeshgm) [inline templates/result-email-template.html:117]: one of the example of `total`  is  `4.0 minutes 0.2220790000000079 seconds` . I this is not ideal for readability. can you reduce the decimal to points ?

File: templates/result-email-template.html [hunk 16]
```diff
@@ -72,13 +98,23 @@ <h2>General Information</h2>
...
-            <td>{{ suite_run_time }}</td>
+            <th style="color: blue">Invoked by User</th>
+            <td>{{ trigger_user }}</td>
+        </tr>
+    </table>
+    <h3>Suite duration</h3>
+    <table class="half-table">
+        <tr>
+            <th style="color: blue">Start Time</th>
+            <td>{{ suite_run_time.start }}</td>
```
Comment (rakeshgm) [inline templates/result-email-template.html:109]: can we have the name representation here instead of number format. This may confuse the geographical lookers :).  And also use 12 hour format here

File: templates/result-email-template.html [hunk 17]
```diff
@@ -72,13 +98,23 @@ <h2>General Information</h2>
...
         </tr>
+        <tr>
+            <th style="color: blue">End Time</th>
+            <td>{{ suite_run_time.end }}</td>
```
Comment (rakeshgm) [inline templates/result-email-template.html:113]: same as above

File: setup.py (hunk 14 above)
Comment (pdhiran) [reply setup.py:24]: There was a conflict with jinja2 Version, where was in a dependent package required for jinja_markdown was not getting installed with current version of jinja2 present. So removing the version solved the dependency issue.
``` ERROR: Cannot install -r requirements.txt (line 1) and cephci because these package versions have conflicting dependencies.

The conflict is caused by:
    cephci 0.1 depends on jinja2==2.10.1
    jinja-markdown  depends on Jinja2~=2.11
```

File: templates/result-email-template.html (hunk 15 above)
Comment (pdhiran) [reply templates/result-email-template.html:117]: Yeah, Sorry, I have changed it since then and the mailw as not with final rendering. I will post the log with all the changes in place.

File: templates/result-email-template.html (hunk 16 above)
Comment (pdhiran) [reply templates/result-email-template.html:109]: Does this format look good? : ` 20 January 2021 at 07:06:20 PM `

Comment (rakeshgm) [reply templates/result-email-template.html:109]: quite good.

**Comment by pdhiran** [comment]
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-1611154763587/

**Comment by pdhiran** [comment]
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-1611157472863/

**Review by psathyan — APPROVED** [review]
Looks good to me

**Review by sunilkumarn417 — DISMISSED** [review]
LGTM

**Comment by rakeshgm** [comment]
@pdhiran the precursor rendering of this template looked nice. i.e having the suite duration as a separate section. you can rearrange the order as @psathyan mentioned.

**Comment by pdhiran** [comment]
Here the render with the latest changes : http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-1611162849965/

**Review by rakeshgm — APPROVED** [review]
